sdktests/server_side_hooks.go

```diff
@@ -0,0 +1,307 @@
+
+	"github.com/launchdarkly/go-server-sdk-evaluation/v3/ldbuilders"
+
+	"github.com/launchdarkly/sdk-test-harness/v2/data"
+	"github.com/launchdarkly/sdk-test-harness/v2/framework/ldtest"
+	o "github.com/launchdarkly/sdk-test-harness/v2/framework/opt"
+	"github.com/launchdarkly/sdk-test-harness/v2/mockld"
+	"github.com/launchdarkly/sdk-test-harness/v2/servicedef"
+
+	"github.com/stretchr/testify/assert"
+
+	"time"
+)
+
+func doServerSideHooksTests(t *ldtest.T) {
+	t.RequireCapability(servicedef.CapabilityEvaluationHooks)
+	t.Run("executes beforeEvaluation stage", executesBeforeEvaluationStage)
+	t.Run("executes afterEvaluation stage", executesAfterEvaluationStage)
+	t.Run("data propagates from before to after", beforeEvaluationDataPropagatesToAfter)
+	t.Run("data propagates from before to after for migrations", beforeEvaluationDataPropagatesToAfterMigration)
+}
+
+func executesBeforeEvaluationStage(t *ldtest.T) {
+	t.Run("without detail", func(t *ldtest.T) { executesBeforeEvaluationStageDetail(t, false) })
+	t.Run("with detail", func(t *ldtest.T) { executesBeforeEvaluationStageDetail(t, true) })
+	t.Run("for migrations", executesBeforeEvaluationStageMigration)
+}
+
+func executesAfterEvaluationStage(t *ldtest.T) {
+	t.Run("without detail", func(t *ldtest.T) { executesAfterEvaluationStageDetail(t, false) })
+	t.Run("with detail", func(t *ldtest.T) { executesAfterEvaluationStageDetail(t, true) })
+	t.Run("for migrations", executesAfterEvaluationStageMigration)
+}
+
+func beforeEvaluationDataPropagatesToAfter(t *ldtest.T) {
+	t.Run("without detail", func(t *ldtest.T) { beforeEvaluationDataPropagatesToAfterDetail(t, false) })
+	t.Run("with detail", func(t *ldtest.T) { beforeEvaluationDataPropagatesToAfterDetail(t, true) })
+}
+
+type VariationParameters struct {
+	name         string
+	flagKey      string
+	defaultValue ldvalue.Value
+	valueType    servicedef.ValueType
+	detail       bool
+}
+
+func variationTestParams(detail bool) []VariationParameters {
+	return []VariationParameters{{
+		name:         "for boolean variation",
+		flagKey:      "bool-flag",
+		defaultValue: ldvalue.Bool(false),
+		valueType:    servicedef.ValueTypeBool,
+		detail:       detail,
+	},
+		{
+			name:         "for string variation",
+			flagKey:      "string-flag",
+			defaultValue: ldvalue.String("default"),
+			valueType:    servicedef.ValueTypeString,
+			detail:       detail,
+		},
+		{
+			name:         "for double variation",
+			flagKey:      "number-flag",
+			defaultValue: ldvalue.Float64(3.14),
+			valueType:    servicedef.ValueTypeDouble,
+			detail:       detail,
+		},
+		{
+			name:         "for int variation",
+			flagKey:      "number-flag",
+			defaultValue: ldvalue.Int(0xDEADBEEF),
+			valueType:    servicedef.ValueTypeInt,
+			detail:       detail,
+		},
+		{
+			name:         "for json variation",
+			flagKey:      "json-flag",
+			defaultValue: ldvalue.ObjectBuild().Build(),
+			valueType:    servicedef.ValueTypeInt,
```

Comment:
Intend for this to be ValueTypeInt?

Reply:
I did not. It also indicates I need to add a value check.

Reply:
Actually. I need to figure out why it passed, because I do have one.

Reply:
Ah, yeah, it doesn't have an independent evaluation. That is probably fine though, this more cares that they match than testing the evaluation itself.